### PR TITLE
feat: Add support for custom named log file names for log-rotate plugin

### PR DESCRIPTION
**Pull request by Revolyssup**
### Description

Supersedes https://github.com/apache/apisix/pull/9210
Fixes https://github.com/apache/apisix/issues/9061
Prior to this PR, the path to access and error log passed in config is not respected and path is hardcoded to logs/error.log and logs/access.log. This PR fixes that as well. 
### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: t/plugin/log-rotate3.t [hunk 1]
```diff
@@ -132,3 +132,40 @@ start xxxxxx
     }
 --- response_body
 passed
+
+
+=== TEST 4: max_kept effective on differently named compression files
```
Comment (monkeyDluffy6017) [inline t/plugin/log-rotate3.t:137]: Could you add a test case with enable_compression = false because the issue you fixed is like this

Comment (Revolyssup) [reply t/plugin/log-rotate3.t:138]: done

**Comment by Revolyssup** [comment]
@monkeyDluffy6017 Please approve CI

File: apisix/core/config_local.lua [hunk 2]
```diff
@@ -55,6 +55,7 @@ end
 -- local local_conf = core.config.local_conf()
 -- local fallback_sni = core.table.try_read_attr(
 --                        local_conf, "apisix", "ssl", "fallback_sni") -- "a.test2.com"
+-- -- `nil` is returned if the attribute doesn't exist
```
Comment (shreemaan-abhishek) [inline apisix/core/config_local.lua:58]: two double dashes?

Comment (shreemaan-abhishek) [inline apisix/core/config_local.lua:58]: two double dashes?

File: t/plugin/log-rotate2.t [hunk 3]
```diff
@@ -181,13 +181,13 @@ plugins:
...
   log-rotate:
-    interval: 1
+    interval: 2
     max_kept: 1
     enable_compression: true
 --- config
     location /t {
         content_by_lua_block {
-            ngx.sleep(3.5)
+            ngx.sleep(2.5)
```
Comment (shreemaan-abhishek) [inline t/plugin/log-rotate2.t:190]: why is this change needed?

File: t/plugin/log-rotate.t [hunk 4]
```diff
@@ -105,6 +106,7 @@ start xxxxxx
 --- config
     location /t {
         content_by_lua_block {
+            local io = require("io")
```
Comment (shreemaan-abhishek) [inline t/plugin/log-rotate.t:109]: I think you will have to revert these changes.

Comment (Revolyssup) [reply t/plugin/log-rotate.t:109]: It fails lint check when I try to use global "io"

Comment (Revolyssup) [reply t/plugin/log-rotate.t:109]: Similar to this one https://github.com/apache/apisix/actions/runs/5218439289/jobs/9419350930?pr=9619#step:4:277

File: apisix/core/config_local.lua (hunk 2 above)
Comment (Revolyssup) [reply apisix/core/config_local.lua:58]: By mistake. I'll fix it

File: t/plugin/log-rotate2.t (hunk 3 above)
Comment (Revolyssup) [reply t/plugin/log-rotate2.t:190]: The tests are flaky and fail intermittently and with this timeout configuration they stopped failing.

File: t/plugin/log-rotate.t (hunk 4 above)
Comment (shreemaan-abhishek) [reply t/plugin/log-rotate.t:109]: You have imported "io" but haven't used it.

Comment (Revolyssup) [reply t/plugin/log-rotate.t:109]: I have used it below in the for loop

Comment (shreemaan-abhishek) [reply t/plugin/log-rotate.t:109]: Strange, I don't see any other additions except importing "io" module in to this file 🤔

**Comment by monkeyDluffy6017** [comment]
@Revolyssup please make the ci happy

**Comment by Revolyssup** [comment]
> @Revolyssup please make the ci happy

The failing tests are unrelated to this change :(

File: apisix/core/config_local.lua (hunk 2 above)
Comment (monkeyDluffy6017) [reply apisix/core/config_local.lua:58]: no need to fix, The second `--` indicates that the line was originally a comment

File: apisix/plugins/log-rotate.lua [hunk 5]
```diff
@@ -72,11 +74,11 @@ local function file_exists(path)
     return file ~= nil
 end
 
-
+-- get the log directory path
```
Comment (monkeyDluffy6017) [inline apisix/plugins/log-rotate.lua:77]: need two blanks

File: apisix/plugins/log-rotate.lua [hunk 6]
```diff
@@ -101,15 +103,15 @@ local function get_log_path_info(file_type)
...
+    return prefix , file_type
 end
 
 
 local function tab_sort_comp(a, b)
     return a > b
 end
 
-
+-- scans the log directory and returns a sorted table of matching log files
```
Comment (monkeyDluffy6017) [inline apisix/plugins/log-rotate.lua:114]: need two blanks

File: apisix/plugins/log-rotate.lua [hunk 7]
```diff
@@ -130,6 +132,18 @@ local function scan_log_folder(log_file_name)
     return t, log_dir
 end
 
+local function get_custom_logfile_name()
```
Comment (monkeyDluffy6017) [inline apisix/plugins/log-rotate.lua:135]: need two blanks

File: apisix/plugins/log-rotate.lua [hunk 8]
```diff
@@ -130,6 +132,18 @@ local function scan_log_folder(log_file_name)
...
+        core.table.try_read_attr(local_conf, "nginx_config", "error_log")
+    local custom_access_log_filename =
+        core.table.try_read_attr(local_conf, "nginx_config", "http", "access_log")
+
+    return custom_access_log_filename, custom_error_log_filename
+
+end
+
 
 local function rename_file(log, date_str)
```
Comment (monkeyDluffy6017) [inline apisix/plugins/log-rotate.lua:148]: need two blanks

File: apisix/plugins/log-rotate.lua [hunk 9]
```diff
@@ -143,10 +157,18 @@ local function rename_file(log, date_str)
         core.log.info("file exist: ", new_file)
         return new_file
     end
-
-    local ok, err = os_rename(log.file, new_file)
```
Comment (monkeyDluffy6017) [inline apisix/plugins/log-rotate.lua:147]: need a blank

File: apisix/plugins/log-rotate.lua [hunk 10]
```diff
@@ -143,10 +157,18 @@ local function rename_file(log, date_str)
...
+    local filename = log.file
+    -- create the file if it does not exist
+    local exists = lfs.attributes(filename, "mode") == "file"
```
Comment (monkeyDluffy6017) [inline apisix/plugins/log-rotate.lua:162]: We have a funciton `file_exists` already

File: apisix/plugins/log-rotate.lua [hunk 11]
```diff
@@ -143,10 +157,18 @@ local function rename_file(log, date_str)
...
+    if not exists then
+        local file = io_open(filename, "w")
+        if file then
+            file:close()
+        end
+    end
+    local ok, err = os_rename(filename, new_file)
```
Comment (monkeyDluffy6017) [inline apisix/plugins/log-rotate.lua:169]: need a blank

File: apisix/plugins/log-rotate.lua [hunk 12]
```diff
@@ -143,10 +157,18 @@ local function rename_file(log, date_str)
...
-    local ok, err = os_rename(log.file, new_file)
+    local filename = log.file
+    -- create the file if it does not exist
+    local exists = lfs.attributes(filename, "mode") == "file"
+    if not exists then
+        local file = io_open(filename, "w")
+        if file then
+            file:close()
+        end
+    end
```
Comment (monkeyDluffy6017) [inline apisix/plugins/log-rotate.lua:168]: Why do you create a new file here, i think it's better to record a error log and exit the function

File: apisix/plugins/log-rotate.lua [hunk 13]
```diff
@@ -46,8 +46,10 @@ local COMPRESSION_FILE_SUFFIX = ".tar.gz" -- compression file suffix
 local rotate_time
 local default_logs
 local enable_compression = false
-local DEFAULT_ACCESS_LOG_FILENAME = "access.log"
-local DEFAULT_ERROR_LOG_FILENAME = "error.log"
+local DEFAULT_ACCESS_LOG_FILENAME = "logs/access.log"
+local DEFAULT_ERROR_LOG_FILENAME = "logs/error.log"
```
Comment (monkeyDluffy6017) [inline apisix/plugins/log-rotate.lua:50]: I think it's better not to change this, the variable name is `FILENAME`, it should not include the file path

File: apisix/plugins/log-rotate.lua [hunk 14]
```diff
@@ -312,6 +336,8 @@ end
 
 function _M.init()
     timers.register_timer("plugin#log-rotate", rotate, true)
+    custom_access_log_filename, custom_error_log_filename = get_custom_logfile_name()
+    timers.register_timer("plugin#log-rotate", rotate, true)
```
Comment (monkeyDluffy6017) [inline apisix/plugins/log-rotate.lua:340]: why do you add one more timer here?

File: apisix/plugins/log-rotate.lua (hunk 12 above)
Comment (monkeyDluffy6017) [reply apisix/plugins/log-rotate.lua:172]: @Revolyssup please check this

File: apisix/plugins/log-rotate.lua (hunk 13 above)
Comment (monkeyDluffy6017) [reply apisix/plugins/log-rotate.lua]: @Revolyssup

File: apisix/plugins/log-rotate.lua (hunk 10 above)
Comment (monkeyDluffy6017) [reply apisix/plugins/log-rotate.lua:166]: @Revolyssup

File: apisix/plugins/log-rotate.lua (hunk 12 above)
Comment (monkeyDluffy6017) [reply apisix/plugins/log-rotate.lua:172]: Is it necessay to add this check?
The original logic looked fine,  `os.rename` would log when it failed and then skip

Comment (Revolyssup) [reply apisix/plugins/log-rotate.lua]: @monkeyDluffy6017 removed this check

**Comment by monkeyDluffy6017** [comment]
Please make the ci pass

File: apisix/plugins/log-rotate.lua [hunk 15]
```diff
@@ -46,8 +46,10 @@ local COMPRESSION_FILE_SUFFIX = ".tar.gz" -- compression file suffix
 local rotate_time
 local default_logs
 local enable_compression = false
-local DEFAULT_ACCESS_LOG_FILENAME = "access.log"
-local DEFAULT_ERROR_LOG_FILENAME = "error.log"
+local DEFAULT_ACCESS_LOG_FILEPATH = "logs/access.log"
+local DEFAULT_ERROR_LOG_FILEPATH = "logs/error.log"
+local custom_access_log_filename
+local custom_error_log_filename
```
Comment (monkeyDluffy6017) [inline apisix/plugins/log-rotate.lua:52]: Should this be `custom_xx_log_filepath` too?

**Comment by monkeyDluffy6017** [comment]
I fix this by https://github.com/apache/apisix/pull/9749

**Comment by monkeyDluffy6017** [comment]
I will close this pr becase it's fixed by #9749
